### ISSUE TITLE
Introduce debug information to statements in MIR

### ISSUE THREAD

**Issue by dianqk** [rust-lang/compiler-team#867]
# Proposal

Currently, the MIR optimization passes are unable to eliminate dead statements that contain debug information. For example:

```rust
struct Foo {
    v1: i32,
    v2: i32
}

#[no_mangle]
fn foo(foo: Foo) {
    let _val = foo.v2; // Because we need the debug information to know the value of `_val` at this point, we cannot eliminate this statement.
}
```

To enable more optimizations within MIR, I propose adding debug information to `Statement`. This will allow us to preserve necessary debug information while eliminating the statements when they are dead.

# Mentors or Reviewers

@wesleywiser (because of your work on MIR optimizations and debug information)

# Process

The main points of the [Major Change Process][MCP] are as follows:

* [x] File an issue describing the proposal.
* [ ] A compiler team member or contributor who is knowledgeable in the area can **second** by writing `@rustbot second`.
    * Finding a "second" suffices for internal changes. If however, you are proposing a new public-facing feature, such as a `-C flag`, then full team check-off is required.
    * Compiler team members can initiate a check-off via `@rfcbot fcp merge` on either the MCP or the PR.
* [ ] Once an MCP is seconded, the Final Comment Period begins. If no objections are raised after 10 days, the MCP is considered **approved**.

You can read [more about Major Change Proposals on forge][MCP].

[MCP]: https://forge.rust-lang.org/compiler/proposals-and-stabilization.html#how-do-i-submit-an-mcp
